### DISCUSSION TITLE
Knowing what ViewModel I was Navigated from

### DISCUSSION THREAD

**Discussion by munkii** [PrismLibrary/Prism#3137] [.NET MAUI] [answered]
What is the "Prism way" of knowing what View\ViewModel I came from? Is there a way of knowing it was Back navigation or Forward

I know I could just roll-my-own solution and in this specfic scenario do something like

```
NavigationParameters np = new NavigationParameters();
np.Add("PreviousViewModel", typeof(BreatheAccountNotConnectedViewModel));
```

But I wanted to know if there was a more Prism correct way of doing it

**Comment by dansiegel** [comment]
```cs
public void OnNavigatedTo(INavigationParameters parameters)
{
    if (parameters.GetNavigationMode() == NavigationMode.New)
    {
        // Will happen the first time the view is navigated to
    }
    else
    {
        // will happen any time the View is navigated back to
    }
}
```

**Reply by munkii** [reply]
Sorry I should have been clearer. If it could be navigated to from two different ViewModels  I'd like to know which one it was.

**Reply by dansiegel** [reply]
Your ViewModel should only care about the state that is either passed to it or that it gets from various services that you've injected, not what navigated to it. Can you elaborate as to why you care where it was navigated from?

**Reply by munkii** [reply]
I have a View\ViewModel that displays the App Account page called BreatheAccountConnectedViewModel. This is navigated to direct from "Settings" page, 99% of the time

When the app is first installed we want to take the user via BreatheAccount**Not**ConnectedViewModel. We sign them in and then send them over to BreatheAccountConnectedViewModel. That journey is on initial login.

When BreatheAccountConnectedViewModel has been navigated from BreatheAccount**Not**ConnectedViewModel I kick off their first ever data sync. In fairness there are probably other ways I could determine if this is the first visit after initial login.

**Reply by dansiegel** [reply, marked as answer]
It would seem to me that this is just a tight coupling and bad design. Your ViewModel should only care about the state that it is either passed via NavigationParameters or via services that you may have.

For example:

```cs
public class SettingsViewModel(IAppSettings settings)
{
    void DoFoo()
    {
        if (settings.AccountConnected)
        {
            // Do something...
        }
        else
        {
            // Do something else...
        }
    }

    public void OnNavigatedTo(INavigationParameters parameters)
    {
        if (paramters.TryGetValue<bool>("accountConnected", out var accountConnected) && accountConnected)
        {
            // Do something...
        }
        else
        {
            // Do something else...
        }
    }
}
```

**Reply by munkii** [reply]
Note passing it as navigationParamater (or at least the FreshMVVM equivalent) is what I am doing at the moment. I just wanted to check there was not some Prism PreviousModel property or somesuch.
